<a href="https://colab.research.google.com/github/XTMay/python-data-science-course/blob/main/notebooks/Lec_3_4_Introduction_to_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 📊 Lesson 03_04: Introduction to [pandas](https://pandas.pydata.org/)

## 🎯 学习目标 / Learning Objectives

	• 了解 pandas 是什么及其在数据科学中的作用
	• 创建 pandas DataFrame 和 Series
	• 学习常用数据操作（索引、切片、筛选）
	• 使用 pandas 进行数据清洗和预处理

# 🔧 什么是 pandas？What is pandas?

- pandas 是 Python 中一个强大的数据分析和处理库。
- 它提供了高性能、易于使用的数据结构，如 DataFrame（二维表格）和 Series（一维数组）。
- pandas 是数据科学家进行数据清洗、转换、分析和可视化的核心工具。

## Create a pandas Series from a Python list

In [ ]:
import pandas as pd

In [ ]:
import pandas as pd

# Create a pandas Series from a Python list
my_list = [10, 20, 30, 40, 50] # List
series_from_list = pd.Series(my_list)
print("Series from list:")
display(series_from_list)

Series from list:


0    10
1    20
2    30
3    40
4    50
dtype: int64

## Create a pandas DataFrame from a Python dictionary

In [ ]:
# Create a pandas DataFrame from a Python dictionary
my_dict = {
    'Column A': [1, 2, 3, 4],
    'Column B': ['A', 'B', 'C', 'D'],
    'Column C': [True, False, True, False]
} # dictionary

dataframe_from_dict = pd.DataFrame(my_dict)
print("\nDataFrame from dictionary:")
display(dataframe_from_dict)


DataFrame from dictionary:


Column A Column B  Column C
0         1        A      True
1         2        B     False
2         3        C      True
3         4        D     False

## Create a pandas DataFrame from a NumPy array

In [ ]:
# Create a pandas DataFrame from a NumPy array
np_array = np.array([
    [100, 200, 300],
    [400, 500, 600],
    [700, 800, 900]
])

dataframe_from_numpy = pd.DataFrame(np_array, columns=['Col1', 'Col2', 'Col3'])
print("\nDataFrame from NumPy array:")
display(dataframe_from_numpy)


DataFrame from NumPy array:


Col1  Col2  Col3
0   100   200   300
1   400   500   600
2   700   800   900

### 🔍 数据查看与基本信息 / Data Inspection and Basic Information

In [ ]:
# Display the first 3 rows
print("First 3 rows using .head():")
display(dataframe_from_dict.head(3))

First 3 rows using .head():


Column A Column B  Column C
0         1        A      True
1         2        B     False
2         3        C      True

In [ ]:
# Display the last 2 rows
print("\nLast 2 rows using .tail():")
display(dataframe_from_dict.tail(2))


Last 2 rows using .tail():


Column A Column B  Column C
2         3        C      True
3         4        D     False

In [ ]:
# Print concise summary of the DataFrame
print("\nDataFrame information using .info():")
dataframe_from_dict.info()


DataFrame information using .info():
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Column A  4 non-null      int64 
 1   Column B  4 non-null      object
 2   Column C  4 non-null      bool  
dtypes: bool(1), int64(1), object(1)
memory usage: 200.0+ bytes


In [ ]:
# Generate descriptive statistics
print("\nDescriptive statistics using .describe():")
display(dataframe_from_dict.describe())


Descriptive statistics using .describe():


Column A
count  4.000000
mean   2.500000
std    1.290994
min    1.000000
25%    1.750000
50%    2.500000
75%    3.250000
max    4.000000

In [ ]:
# Print the shape of the DataFrame
print("\nShape of the DataFrame using .shape:")
print(dataframe_from_dict.shape)


Shape of the DataFrame using .shape:
(4, 3)


In [ ]:
# Print the column labels
print("\nColumn labels using .columns:")
print(dataframe_from_dict.columns)


Column labels using .columns:
Index(['Column A', 'Column B', 'Column C'], dtype='object')


### 🔍 索引与选择 / Indexing and Selection


In [ ]:
# Demonstrate .loc for label-based indexing
print("Original DataFrame:")
display(dataframe_from_dict)

Original DataFrame:


Column A Column B  Column C
0         1        A      True
1         2        B     False
2         3        C      True
3         4        D     False

In [ ]:
print("\nSelect element at row label 0, column label 'Column B':")
display(dataframe_from_dict.loc[0, 'Column B']) # location


Select element at row label 0, column label 'Column B':


'A'

In [ ]:
print("\nSelect row with label 2:")
display(dataframe_from_dict.loc[2])


Select row with label 2:


Column A       3
Column B       C
Column C    True
Name: 2, dtype: object

In [ ]:
# Demonstrate .loc for label-based indexing
print("Original DataFrame:")
display(dataframe_from_dict)

Original DataFrame:


Column A Column B  Column C
0         1        A      True
1         2        B     False
2         3        C      True
3         4        D     False

## index_a : index_b (Range):  index_a <= i < index_b

# loc(param_row, param_col)

In [ ]:
print("\nSelect column with label 'Column A':")
display(dataframe_from_dict.loc[1:3, 'Column C']) # index_a : index_b:  index_a <= i <= index_b


Select column with label 'Column A':


1    False
2     True
3    False
Name: Column C, dtype: bool

In [ ]:
print("Original DataFrame:")
display(dataframe_from_dict)

Original DataFrame:


Column A Column B  Column C
0         1        A      True
1         2        B     False
2         3        C      True
3         4        D     False

# dataframe_from_dict.loc[1:3, 'Column B':'Column C']

In [ ]:
print("\nSelect rows with labels 1 to 3 (inclusive) and columns 'Column B' to 'Column C':")
display(dataframe_from_dict.loc[1:3, 'Column B':'Column C'])


Select rows with labels 1 to 3 (inclusive) and columns 'Column B' to 'Column C':


Column B  Column C
1        B     False
2        C      True
3        D     False

In [ ]:
# Demonstrate boolean indexing
print("Original DataFrame:")
display(dataframe_from_dict)

Original DataFrame:


Column A Column B  Column C
0         1        A      True
1         2        B     False
2         3        C      True
3         4        D     False

# dataframe[dataframe['Column'] == condition?]

In [ ]:
print("\nSelect rows where 'Column C' is True:")
display(dataframe_from_dict[dataframe_from_dict['Column C'] == False])


Select rows where 'Column C' is True:


Column A Column B  Column C
1         2        B     False
3         4        D     False

In [ ]:
# Demonstrate boolean indexing
print("Original DataFrame:")
display(dataframe_from_dict)

Original DataFrame:


Column A Column B  Column C
0         1        A      True
1         2        B     False
2         3        C      True
3         4        D     False

# col_a >= 2
# col_c = True

dataframe_from_dict[(dataframe_from_dict['Column A'] >= 2) & (dataframe_from_dict['Column C'] == True)]

In [ ]:
print("\nSelect rows where 'Column A' is greater than 2:")
display(dataframe_from_dict[dataframe_from_dict['Column A'] > 2])


Select rows where 'Column A' is greater than 2:


Column A Column B  Column C
2         3        C      True
3         4        D     False

# AND: &  , OR: |
# (condition_1) & (condition_2)

In [ ]:
print("\nSelect rows where 'Column B' is 'A' or 'D':")
display(dataframe_from_dict[(dataframe_from_dict['Column B'] == 'A') | (dataframe_from_dict['Column B'] == 'D')])

### 🧼 处理缺失数据 / Handling Missing Data

In [ ]:
# Create a DataFrame with missing values
data_with_missing = {
    'A': [1, 2, np.nan, 4, 5],
    'B': ['a', 'b', 'c', np.nan, 'e'],
    'C': [True, False, True, False, np.nan],
    'D': [10.5, np.nan, 20.1, 30.0, 40.2]
}
df_missing = pd.DataFrame(data_with_missing)
print("Original DataFrame with missing values:")
display(df_missing)

Original DataFrame with missing values:


A    B      C     D
0  1.0    a   True  10.5
1  2.0    b  False   NaN
2  NaN    c   True  20.1
3  4.0  NaN  False  30.0
4  5.0    e    NaN  40.2

# isnull()

In [ ]:
# Identify missing values using .isnull()
print("\nIdentifying missing values using .isnull():")
display(df_missing.isnull())


Identifying missing values using .isnull():


A      B      C      D
0  False  False  False  False
1  False  False  False   True
2   True  False  False  False
3  False   True  False  False
4  False  False   True  False

# Fill missing values using .fillna()

In [ ]:
# Fill missing values using .fillna()

# Fill with a specific value (e.g., 0 for numeric, 'Unknown' for string)
print("\nFilling missing numeric values with 0:")
df_filled_zero = df_missing.fillna(99)
display(df_filled_zero)


Filling missing numeric values with 0:


A   B      C     D
0   1.0   a   True  10.5
1   2.0   b  False  99.0
2  99.0   c   True  20.1
3   4.0  99  False  30.0
4   5.0   e     99  40.2

# fillna({'B': 'Unknown', 'C': False})

In [ ]:
print("\nFilling missing string values with 'Unknown':")
df_filled_unknown = df_missing.fillna({'A': 3, 'D': 30, 'B': "f"}) # Fill specific columns
display(df_filled_unknown)


Filling missing string values with 'Unknown':


A  B      C     D
0  1.0  a   True  10.5
1  2.0  b  False  30.0
2  3.0  c   True  20.1
3  4.0  f  False  30.0
4  5.0  e    NaN  40.2

# fillna(df_filled_mean['A'].mean())

In [ ]:
# Fill with the mean of the column (only works for numeric columns)
print("\nFilling missing numeric values with column mean:")
df_filled_mean = df_missing.copy() # Create a copy to avoid modifying original
df_filled_mean['A'] = df_filled_mean['A'].fillna(df_filled_mean['A'].mean())
df_filled_mean['D'] = df_filled_mean['D'].fillna(df_filled_mean['D'].mean())
display(df_filled_mean)


Filling missing numeric values with column mean:


A    B      C     D
0  1.0    a   True  10.5
1  2.0    b  False  25.2
2  3.0    c   True  20.1
3  4.0  NaN  False  30.0
4  5.0    e    NaN  40.2

# dropna()

In [ ]:
# Drop rows or columns with missing values using .dropna()
# Drop rows with any missing values (default: axis=0)
print("\nDropping rows with any missing values:")
df_dropped_rows = df_missing.dropna()
display(df_dropped_rows)


Dropping rows with any missing values:


A  B     C     D
0  1.0  a  True  10.5

# dropna()

# dropna(axis=1)

In [ ]:
# Drop columns with any missing values (axis=1)
print("\nDropping columns with any missing values:")
df_dropped_cols = df_missing.dropna(axis=1)
display(df_dropped_cols)


Dropping columns with any missing values:


D
0  10.5
1  15.5
2  20.1
3  30.0
4  40.2

In [ ]:
# Create a DataFrame with missing values
data_with_missing = {
    'A': [np.nan, np.nan, np.nan, np.nan, np.nan],
    'B': ['a', 'b', np.nan, np.nan, 'e'],
    'C': [True, False, np.nan, False, np.nan],
    'D': [10.5, np.nan, np.nan, 30.0, 40.2]
}
df_missing = pd.DataFrame(data_with_missing)
print("Original DataFrame with missing values:")
display(df_missing)

Original DataFrame with missing values:


A    B      C     D
0 NaN    a   True  10.5
1 NaN    b  False   NaN
2 NaN  NaN    NaN   NaN
3 NaN  NaN  False  30.0
4 NaN    e    NaN  40.2

In [ ]:
# Drop rows only if all values are missing
print("\nDropping rows only if all values are missing (using how='all'):")
df_dropped_all_rows = df_missing.dropna(how='all')
display(df_dropped_all_rows)


Dropping rows only if all values are missing (using how='all'):


A    B      C     D
0  1.0    a   True  10.5
1  2.0    b  False   NaN
3  4.0  NaN  False  30.0
4  5.0    e    NaN  40.2

# dropna(axis=1, how='any')

In [ ]:
# Drop columns only if all values are missing
print("\nDropping columns only if all values are missing (using how='all'):")
df_dropped_all_cols = df_missing.dropna(axis=1, how='all')
display(df_dropped_all_cols)


Dropping columns only if all values are missing (using how='all'):


B      C     D
0    a   True  10.5
1    b  False   NaN
2  NaN    NaN   NaN
3  NaN  False  30.0
4    e    NaN  40.2

### ➕ 基本运算 / Basic Operations


In [ ]:
# Define two pandas Series
series1 = pd.Series([10, 20, 30, 40, 50])
series2 = pd.Series([1, 2, 3, 4, 5])

print("Series 1:")
display(series1)
print("\nSeries 2:")
display(series2)

Series 1:


0    10
1    20
2    30
3    40
4    50
dtype: int64


Series 2:


0    1
1    2
2    3
3    4
4    5
dtype: int64

In [ ]:
# Demonstrate element-wise operations
print("\nAddition (series1 + series2):")
display(series1 + series2)


Addition (series1 + series2):


0    11.0
1    22.0
2    33.0
3    44.0
4     NaN
dtype: float64

In [ ]:
print("\nSubtraction (series1 - series2):")
display(series1 - series2)


Subtraction (series1 - series2):


0     9
1    18
2    27
3    36
4    45
dtype: int64

In [ ]:
print("\nMultiplication (series1 * series2):")
display(series1 * series2)


Multiplication (series1 * series2):


0     10
1     40
2     90
3    160
4    250
dtype: int64

In [ ]:
print("\nDivision (series1 / series2):")
display(series1 / series2)


Division (series1 / series2):


0    10.0
1    10.0
2    10.0
3    10.0
4    10.0
dtype: float64

In [ ]:
# Define a simple pandas DataFrame
df_ops = pd.DataFrame({
    'A': [1, 2, 3, 4, 5],
    'B': [10, 20, 30, 40, 50],
    'C': [100, 200, 300, 400, 500]
})

print("Original DataFrame:")
display(df_ops)


Original DataFrame:


A   B    C
0  1  10  100
1  2  20  200
2  3  30  300
3  4  40  400
4  5  50  500

In [ ]:
# Demonstrate basic statistical functions
print("\nSum of all elements in the DataFrame:")
display(df_ops.sum()) # Sum along columns (axis=0 by default)


Sum of all elements in the DataFrame:


A      15
B     150
C    1500
dtype: int64

In [ ]:
print("\nMean of column 'A':")
display(df_ops['A'].mean())


Mean of column 'A':


np.float64(3.0)

In [ ]:
print("\nMaximum value in column 'B':")
display(df_ops['B'].max())


Maximum value in column 'B':


50

In [ ]:
print("\nMinimum value in the entire DataFrame:")
display(df_ops.min()) # Minimum along columns (axis=0 by default)


Minimum value in the entire DataFrame:


A      1
B     10
C    100
dtype: int64

In [ ]:
df_ops

A   B    C
0  1  10  100
1  2  20  200
2  3  30  300
3  4  40  400
4  5  50  500

In [ ]:
print("\nSum of rows:")
display(df_ops.sum(axis=1)) # Sum along rows (axis=1)


Sum of rows:


0    111
1    222
2    333
3    444
4    555
dtype: int64

In [ ]:
# Demonstrate applying mathematical operations

print("Original Series (series1):")
display(series1)

Original Series (series1):


0    10
1    20
2    30
3    40
4    50
dtype: int64

In [ ]:

print("\nSeries 1 squared (series1 ** 2):")
display(series1 ** 2)



Series 1 squared (series1 ** 2):


0     100
1     400
2     900
3    1600
4    2500
dtype: int64

In [ ]:
import numpy as np

In [ ]:
print("\nSquare root of Series 1 (np.sqrt(series1)):")
display(np.sqrt(series1))


Square root of Series 1 (np.sqrt(series1)):


0    3.162278
1    4.472136
2    5.477226
3    6.324555
4    7.071068
dtype: float64

In [ ]:
print("\nOriginal DataFrame (df_ops):")
display(df_ops)


Original DataFrame (df_ops):


A   B    C
0  1  10  100
1  2  20  200
2  3  30  300
3  4  40  400
4  5  50  500

In [ ]:
print("\nDataFrame squared (df_ops ** 2):")
display(df_ops ** 2)


DataFrame squared (df_ops ** 2):


A     B       C
0   1   100   10000
1   4   400   40000
2   9   900   90000
3  16  1600  160000
4  25  2500  250000

In [ ]:
print("\nSquare root of DataFrame (np.sqrt(df_ops)):")
display(np.sqrt(df_ops))


Square root of DataFrame (np.sqrt(df_ops)):


A         B          C
0  1.000000  3.162278  10.000000
1  1.414214  4.472136  14.142136
2  1.732051  5.477226  17.320508
3  2.000000  6.324555  20.000000
4  2.236068  7.071068  22.360680

### 📊 分组与聚合 / Grouping and Aggregation


In [ ]:
# Create a sample DataFrame
data = {
    'Category': ['A', 'B', 'A', 'C', 'B', 'C', 'A', 'B', 'C', 'A'],
    'Value': [10, 15, 12, 18, 20, 25, 11, 16, 22, 13],
    'Another_Value': [100, 150, 120, 180, 200, 250, 110, 160, 220, 130]
}
df_group = pd.DataFrame(data)

print("Original DataFrame:")
display(df_group)

Original DataFrame:


Category  Value  Another_Value
0        A     10            100
1        B     15            150
2        A     12            120
3        C     18            180
4        B     20            200
5        C     25            250
6        A     11            110
7        B     16            160
8        C     22            220
9        A     13            130

In [ ]:
# Group by 'Category' and calculate the sum of 'Value' for each group
print("\nSum of 'Value' grouped by 'Category':")
grouped_sum = df_group.groupby('Category')['Value'].max()
display(grouped_sum)


Sum of 'Value' grouped by 'Category':


Category
A    13
B    20
C    25
Name: Value, dtype: int64

In [ ]:
# Demonstrate multiple aggregation functions
print("\nMultiple aggregations on 'Value' and 'Another_Value' grouped by 'Category':")
grouped_multi_agg = df_group.groupby('Category').agg({
    'Value': ['sum', 'mean', 'count'],
    'Another_Value': ['min', 'max']
})
display(grouped_multi_agg)


Multiple aggregations on 'Value' and 'Another_Value' grouped by 'Category':


Value                  Another_Value     
           sum       mean count           min  max
Category                                          
A           46  11.500000     4           100  130
B           51  17.000000     3           150  200
C           65  21.666667     3           180  250

In [ ]:
# Create a DataFrame with multiple grouping columns
data_multi_group = {
    'City': ['Tokyo', 'Tokyo', 'Osaka', 'Osaka', 'Tokyo', 'Osaka'],
    'Product': ['A', 'B', 'A', 'B', 'A', 'B'],
    'Item': ['I1', 'I1', 'I2', 'I2', 'I1', 'I2'],
    'Sales': [100, 150, 120, 180, 110, 200]
}
df_multi_group = pd.DataFrame(data_multi_group)

print("\nDataFrame for multi-level grouping:")
display(df_multi_group)


DataFrame for multi-level grouping:


City Product Item  Sales
0  Tokyo       A   I1    100
1  Tokyo       B   I1    150
2  Osaka       A   I2    120
3  Osaka       B   I2    180
4  Tokyo       A   I1    110
5  Osaka       B   I2    200

Group: Item

Resutl: City: Count, Product: Count, Sales: sum, max, min

In [ ]:
# Demonstrate multiple aggregation functions
print("\nMultiple aggregations on 'Value' and 'Another_Value' grouped by 'Category':")
grouped_by_city_product = df_multi_group.groupby('Item').agg({
    'City': ['count'],
    'Product': ['count'],
    'Sales': ['sum', 'max', 'min']
})
display(grouped_by_city_product)


Multiple aggregations on 'Value' and 'Another_Value' grouped by 'Category':


City Product Sales          
     count   count   sum  max  min
Item                              
I1       3       3   360  150  100
I2       3       3   500  200  120

In [ ]:
# Group by multiple columns and apply aggregation
print("\nSum of 'Sales' grouped by 'City' and 'Product':")
grouped_by_city_product = df_multi_group.groupby(['Item'])['Sales'].max()
display(grouped_by_city_product)


Sum of 'Sales' grouped by 'City' and 'Product':


Item
I1    150
I2    200
Name: Sales, dtype: int64

## 🤝 合并与连接 DataFrame / Merging and Joining DataFrames

In [ ]:
# Perform an inner merge
print("\nInner Merge:")
inner_merge = pd.merge(df1, df2, on='key', how='inner')
display(inner_merge)


Inner Merge:


key  value1  value2
0   B       2       5
1   D       4       6

In [ ]:
# Create sample DataFrames for merging
df1 = pd.DataFrame({
    'key': ['A', 'B', 'C', 'D'],
    'value1': [1, 2, 3, 4]
})

df2 = pd.DataFrame({
    'key': ['B', 'D', 'E', 'F'],
    'value2': [5, 6, 7, 8]
})

print("DataFrame 1:")
display(df1)
print("\nDataFrame 2:")
display(df2)

DataFrame 1:


key  value1
0   A       1
1   B       2
2   C       3
3   D       4


DataFrame 2:


key  value2
0   B       5
1   D       6
2   E       7
3   F       8

In [ ]:
# Perform a left merge
print("\nLeft Merge:")
left_merge = pd.merge(df2, df1, on='key', how='left')
display(left_merge)


Left Merge:


key  value2  value1
0   B       5     2.0
1   D       6     4.0
2   E       7     NaN
3   F       8     NaN

In [ ]:
# Create sample DataFrames for joining (using index)
df3 = pd.DataFrame({
    'value3': [10, 20, 30]
}, index=['A', 'B', 'C'])

df4 = pd.DataFrame({
    'value4': [40, 50, 60]
}, index=['B', 'C', 'D'])

print("\nDataFrame 3:")
display(df3)
print("\nDataFrame 4:")
display(df4)


DataFrame 3:


value3
A      10
B      20
C      30


DataFrame 4:


value4
B      40
C      50
D      60

In [ ]:
# Perform a left join using the .join() method
print("\nLeft Join (df3.join(df4)):")
left_join = df3.join(df4, how='left')
display(left_join)


Left Join (df3.join(df4)):


value3  value4
A      10     NaN
B      20    40.0
C      30    50.0

### 🧠 练习题 / Practice Exercises

### ✅ 练习 1：数组创建与统计分析

 **题目：**

 创建一个包含 100 个介于 10 和 50 之间的整数的随机 Series，完成以下任务：

     •   输出最大值、最小值、平均值、中位数
     •   统计每个数字出现的次数（提示：value_counts()）

In [ ]:
# 在这里写代码👇

# 创建随机 Series

# 计算统计指标

# 统计每个数字出现频率

 ### ✅ 练习 2：处理缺失数据

 **题目：** 给定一个包含缺失值的 DataFrame。

 请完成以下操作：  

     •   找出每列的缺失值数     
     •   使用该列的平均值填充数值列的缺失值     
     •   使用众数填充非数值列的缺失值     
     •   删除所有包含缺失值的行

In [ ]:

# 在这里写代码👇

# 创建包含缺失值的 DataFrame
data_missing_ex2 = {
    'A': [1, 2, np.nan, 4, 5],
    'B': ['apple', 'banana', np.nan, 'apple', 'orange'],
    'C': [10.5, 11.2, 10.5, np.nan, 12.0],
    'D': [True, False, np.nan, True, False]
}
df_missing_ex2 = pd.DataFrame(data_missing_ex2)

# 找出每列的缺失值数量

# 使用平均值填充数值列

# 使用众数填充非数值列

# 删除包含缺失值的行

### ✅ 练习 3：分组与聚合
 **题目：**

 给定一个销售数据集 DataFrame。请完成以下操作：

     •   按 'Product' 分组，计算每个产品的总销售额
     •   按 'Region' 和 'Product' 分组，计算每个区域和产品的平均销售额
     •   按 'Region' 分组，计算每个区域的总销售额和销售数量

In [ ]:
# 在这里写代码👇

# 创建销售数据集 DataFrame
data_sales_ex3 = {
    'Region': ['North', 'South', 'North', 'East', 'South', 'West', 'North', 'East', 'South', 'West'],
    'Product': ['A', 'B', 'A', 'C', 'B', 'A', 'C', 'B', 'C', 'A'],
    'Sales': [100, 150, 120, 180, 200, 110, 130, 160, 220, 140]
}
df_sales_ex3 = pd.DataFrame(data_sales_ex3)


# 按 'Product' 分组计算总销售额

# 按 'Region' 和 'Product' 分组计算平均销售额

# 按 'Region' 分组计算总销售额和销售数量

### ✅ 总结 / Summary




| #  | pandas 概念/操作                      | 说明                                                                                 |
|----|--------------------------------------|--------------------------------------------------------------------------------------|
| 1  | `DataFrame`                          | 二维带标签的数据结构，类似表格，每列可以是不同的数据类型                            |
| 2  | `Series`                             | 一维带标签的数据结构，类似数组，可看作是 DataFrame 的一列或一行                      |
| 3  | 数据查看 (`head`, `tail`)             | 显示 DataFrame 的前几行（`head()`）或后几行（`tail()`）                              |
| 4  | 数据结构信息 (`info`, `shape`)        | 查看数据类型、非空值数量、行列数等结构信息                                          |
| 5  | 数据摘要 (`describe`)                 | 查看数值列的统计摘要（平均数、标准差、最小值、四分位数等）                          |
| 6  | 访问列 (`df['col']`, `df.col`)       | 通过列名访问 DataFrame 的列                                                         |
| 7  | 访问行 (`.loc`, `.iloc`)              | 使用标签（`.loc`）或位置（`.iloc`）选择指定的行/列                                   |
| 8  | 条件选择（布尔索引）                  | 通过条件筛选数据，如：`df[df['age'] > 30]`                                          |
| 9  | 缺失值处理 (`isnull`, `fillna`, `dropna`) | 识别缺失值（NaN）、填充缺失值或删除含缺失值的行/列                                   |
| 10 | 重命名 (`rename`)                    | 重命名列或索引名，如：`df.rename(columns={'old':'new'})`                            |
| 11 | 排序 (`sort_values`, `sort_index`)   | 根据列值或索引排序                                                                  |
| 12 | 类型转换 (`astype`)                  | 修改列的数据类型，如：`df['col'].astype(int)`                                       |
| 13 | 数据替换 (`replace`)                 | 替换某些值，如：`df['gender'].replace({'M': 'Male'})`                               |
| 14 | 重置索引 (`reset_index`)             | 将索引还原为普通列，常用于整理数据后重新编号                                        |
| 15 | 设置索引 (`set_index`)               | 指定某一列作为新的索引列                                                            |
| 16 | 基本运算（加减乘除、平均、求和等）    | 对整个 DataFrame 或 Series 进行数学和统计运算，如：`df.mean()`, `df.sum()`          |
| 17 | 分组与聚合 (`groupby`, `agg`)        | 按某列分组，并对组内数据进行聚合分析                                                |
| 18 | 透视表 (`pivot_table`)               | 创建类似 Excel 透视表的多维度聚合数据表                                             |
| 19 | 合并与连接 (`merge`, `join`, `concat`) | 多个 DataFrame 的合并连接（横向/纵向）                                              |
| 20 | 应用函数 (`apply`, `map`, `applymap`) | 对行、列、元素级别使用自定义函数                                                    |
| 21 | 时间序列处理                         | 处理时间数据，如：`pd.to_datetime()`, `resample()` 等                               |
| 22 | 数据导入导出（CSV/Excel/JSON等）      | 使用 `read_csv`, `to_excel`, `read_json` 等读取与保存数据                            |

# Netflix Dataset Analysis Example

## Load the datasets
Load the `Netflix_Dataset_Movie.csv` and `Netflix_Dataset_Rating.csv` files into pandas DataFrames.


In [ ]:
import pandas as pd

df_movies = pd.read_csv('/content/drive/MyDrive/AI_Lecture/AI_Data_Scientist/dataset/Netflix_Dataset_Movie.csv')
df_ratings = pd.read_csv('/content/drive/MyDrive/AI_Lecture/AI_Data_Scientist/dataset/Netflix_Dataset_Rating.csv')

In [ ]:
df_movies.head()

Movie_ID  Year                          Name
0         1  2003               Dinosaur Planet
1         2  2004    Isle of Man TT 2004 Review
2         3  1997                     Character
3         4  1994  Paula Abdul's Get Up & Dance
4         5  2004      The Rise and Fall of ECW

In [ ]:
df_ratings.head()

User_ID  Rating  Movie_ID
0   712664       5         3
1  1331154       4         3
2  2632461       3         3
3    44937       5         3
4   656399       4         3

## Inspect the data

Display the first few rows, check the data types, and look for missing values in both DataFrames.


**Reasoning**:
Display the first few rows of both dataframes, check their data types, and count missing values for initial data inspection.



In [ ]:
print("First 5 rows of df_movies:")
display(df_movies.head())

First 5 rows of df_movies:


Movie_ID  Year                          Name
0         1  2003               Dinosaur Planet
1         2  2004    Isle of Man TT 2004 Review
2         3  1997                     Character
3         4  1994  Paula Abdul's Get Up & Dance
4         5  2004      The Rise and Fall of ECW

In [ ]:
print("\nFirst 5 rows of df_ratings:")
display(df_ratings.head())


First 5 rows of df_ratings:


User_ID  Rating  Movie_ID
0   712664       5         3
1  1331154       4         3
2  2632461       3         3
3    44937       5         3
4   656399       4         3

In [ ]:
print("\nInfo of df_movies:")
df_movies.info()


Info of df_movies:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17770 entries, 0 to 17769
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Movie_ID  17770 non-null  int64 
 1   Year      17770 non-null  int64 
 2   Name      17770 non-null  object
dtypes: int64(2), object(1)
memory usage: 416.6+ KB


In [ ]:
print("\nInfo of df_ratings:")
df_ratings.info()


Info of df_ratings:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17337458 entries, 0 to 17337457
Data columns (total 3 columns):
 #   Column    Dtype
---  ------    -----
 0   User_ID   int64
 1   Rating    int64
 2   Movie_ID  int64
dtypes: int64(3)
memory usage: 396.8 MB


In [ ]:
print("\nMissing values in df_movies:")
display(df_movies.isnull().sum())


Missing values in df_movies:


Movie_ID    0
Year        0
Name        0
dtype: int64

In [ ]:
print("\nMissing values in df_ratings:")
display(df_ratings.isnull().sum())


Missing values in df_ratings:


User_ID     0
Rating      0
Movie_ID    0
dtype: int64

## Merge the datasets

Combine the two DataFrames based on a common column (likely a movie ID).


# user_location, user_age, user_gender...

In [ ]:
df_merged = pd.merge(df_ratings, df_movies, on='Movie_ID', how='inner')
print("First 5 rows of the merged DataFrame:")
display(df_merged.head())

First 5 rows of the merged DataFrame:


User_ID  Rating  Movie_ID  Year       Name
0   712664       5         3  1997  Character
1  1331154       4         3  1997  Character
2  2632461       3         3  1997  Character
3    44937       5         3  1997  Character
4   656399       4         3  1997  Character

## Analyze movie ratings

Calculate the average rating for each movie and find the movies with the highest and lowest average ratings.


In [ ]:
# Group by 'Movie_ID' and calculate the mean of 'Rating'
average_ratings = df_merged.groupby('Movie_ID')['Rating'].mean()

# Sort in ascending order to find lowest average ratings
lowest_rated_movies = average_ratings.sort_values(ascending=True)

# Sort in descending order to find highest average ratings
highest_rated_movies = average_ratings.sort_values(ascending=False)

# Display the movies with the highest and lowest average ratings
print("Movies with the lowest average ratings:")
display(lowest_rated_movies.head())

Movies with the lowest average ratings:


Movie_ID
3021    1.962031
2276    2.101550
4255    2.203597
3505    2.217361
749     2.221938
Name: Rating, dtype: float64

In [ ]:
print("\nMovies with the highest average ratings:")
display(highest_rated_movies.head())


Movies with the highest average ratings:


Movie_ID
3456    4.665432
2102    4.589824
3444    4.520766
1476    4.461601
4238    4.457774
Name: Rating, dtype: float64

## Analyze user behavior

Determine which users have rated the most movies.


In [ ]:
user_rating_counts

User_ID
6          145
7          192
79         189
97         112
134        166
          ... 
2649370     62
2649378     69
2649388     76
2649426     74
2649429     62
Length: 143458, dtype: int64

In [ ]:
user_rating_counts = df_merged.groupby('User_ID').size()
top_users = user_rating_counts.sort_values(ascending=False)
print("Top users by number of ratings:")
display(top_users.head())

Top users by number of ratings:


User_ID
305344     1344
387418     1339
2439493    1324
2118461    1305
1664010    1257
dtype: int64

## Analyze movie popularity

Find the most frequently rated movies.


In [ ]:
movie_rating_counts

Movie_ID
3        1524
8        9379
16       2517
17       4165
18       9419
        ...  
4488    27997
4490     7507
4492     7511
4493     4743
4496     7835
Length: 1350, dtype: int64

In [ ]:
most_frequently_rated_movies

Movie_ID
1905    117075
2452    102721
4306    102376
571     101450
3860     98545
         ...  
4161      1215
1375      1213
717       1212
2870      1092
4238      1042
Length: 1350, dtype: int64

In [ ]:
# Group by 'Movie_ID' and count the number of ratings
movie_rating_counts = df_merged.groupby('Movie_ID').size()

# Sort the counts in descending order
most_frequently_rated_movies = movie_rating_counts.sort_values(ascending=False)

# Display the top 5 most frequently rated movies
print("Top 5 most frequently rated movies:")
display(most_frequently_rated_movies.head())

Top 5 most frequently rated movies:


Movie_ID
1905    117075
2452    102721
4306    102376
571     101450
3860     98545
dtype: int64

## Summary:

### Data Analysis Key Findings

*   The `df_movies` DataFrame contains 17,770 movies with columns for `Movie_ID`, `Year`, and `Name`, and no missing values.
*   The `df_ratings` DataFrame contains 17,337,458 ratings with columns for `User_ID`, `Rating`, and `Movie_ID`, and no missing values.
*   The datasets were successfully merged based on `Movie_ID` to create `df_merged`.
*   The average rating for each movie was calculated, and the movies with the highest and lowest average ratings were identified.
*   The users who have rated the most movies were identified by counting the number of ratings per user.
*   The most frequently rated movies were identified by counting the number of ratings per movie.

### Insights or Next Steps

*   Further analysis could involve joining the average rating and rating count information back to the movie details to display movie names alongside their popularity and average rating.
*   Consider exploring the distribution of ratings or analyzing user rating patterns to gain deeper insights into user preferences.
